controllers/cluster_controller.go

```diff
@@ -18,34 +18,230 @@ package controllers
+		cluster.SetGroupVersionKind(gvk)
+		if err := r.Client.Status().Patch(ctx, cluster, patchCluster); err != nil {
+			klog.Errorf("Error Patching Cluster status %q in namespace %q: %v", cluster.Name, cluster.Namespace, err)
+			if reterr == nil {
+				reterr = err
+			}
+		}
+	}()
+
+	// If object hasn't been deleted and doesn't have a finalizer, add one.
+	if cluster.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !util.Contains(cluster.Finalizers, clusterv1.ClusterFinalizer) {
+			cluster.Finalizers = append(cluster.ObjectMeta.Finalizers, clusterv1.ClusterFinalizer)
+		}
+	}
+
+	if err := r.reconcile(ctx, cluster); err != nil {
+		if requeueErr, ok := errors.Cause(err).(capierrors.HasRequeueAfterError); ok {
+			klog.Infof("Reconciliation for Cluster %q in namespace %q asked to requeue: %v", cluster.Name, cluster.Namespace, err)
+			return ctrl.Result{Requeue: true, RequeueAfter: requeueErr.GetRequeueAfter()}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	if !cluster.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, cluster)
+	}
 
 	return ctrl.Result{}, nil
 }
 
-func (r *ClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&clusterv1alpha2.Cluster{}).
-		Complete(r)
+// reconcileDelete handles cluster deletion.
+// TODO(ncdc): consolidate all deletion logic in here.
+func (r *ClusterReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster) (reconcile.Result, error) {
+	children, err := r.listChildren(ctx, cluster)
+	if err != nil {
+		klog.Errorf("Failed to list children of cluster %s/%s: %v", cluster.Namespace, cluster.Name, err)
+		return reconcile.Result{}, err
+	}
+
+	if len(children) > 0 {
+		klog.Infof("Cluster %s/%s still has %d children - deleting them first", cluster.Namespace, cluster.Name, len(children))
+
+		var errs []error
+
+		for _, child := range children {
+			accessor, err := meta.Accessor(child)
+			if err != nil {
+				klog.Errorf("Cluster %s/%s: couldn't create accessor for type %T: %v", cluster.Namespace, cluster.Name, child, err)
+				continue
+			}
+
+			if !accessor.GetDeletionTimestamp().IsZero() {
+				// Don't handle deleted child
+				continue
+			}
+
+			gvk := child.GetObjectKind().GroupVersionKind().String()
+
+			klog.Infof("Cluster %s/%s: deleting child %s %s", cluster.Namespace, cluster.Name, gvk, accessor.GetName())
+			if err := r.Delete(context.Background(), child); err != nil {
+				err = errors.Wrapf(err, "error deleting cluster %s/%s: failed to delete %s %s", cluster.Namespace, cluster.Name, gvk, accessor.GetName())
+				klog.Errorf(err.Error())
+				errs = append(errs, err)
+			}
+		}
+
+		if len(errs) > 0 {
+			return reconcile.Result{}, kerrors.NewAggregate(errs)
+		}
+
+		// Requeue so we can check the next time to see if there are still any children left.
+		return reconcile.Result{RequeueAfter: deleteRequeueAfter}, nil
```

Comment:
is there a reason you're not just watching children here?  `.For(&Cluster{}).Owns(&MachineDeployment{})` etc.

Reply:
That's a good point, I think this is because we delete children when we're deleting this Cluster object. @ncdc will know more.